### ISSUE TITLE
LUKS KDF: Switch to argon2id

### ISSUE THREAD

**Issue by no-usernames-left** [QubesOS/qubes-issues#8582]
### The problem you're addressing (if any)

QubesOS 4.1 seems to use argon2i (**not** argon2id) by default for the LUKS KDF.

It would appear as though argon2i is less resistant to attack by sophisticated adversaries than initially thought ([1](https://old.reddit.com/r/linux/comments/12q51ce/comment/jgpvsqc/),[2](https://mjg59.dreamwidth.org/66429.html)).

### The solution you'd like

QubesOS should switch to using argon2id for all new installs, period.

Ideally, existing installs should have their KDF converted at their next QubesOS update (`sudo cryptsetup luksConvertKey /dev/whatever --pbkdf argon2id` etc).

Ideally, [memory cost should be 1024 MiB, ops cost 4, parallelism 1](https://soatok.blog/2022/12/29/what-we-do-in-the-etc-shadow-cryptography-with-passwords/).

### The value to a user, and who that user might be

This will help protect all Qubes users against compromise of their data at rest by sophisticated adversaries.

**Comment by ldeso** [comment]
In my experience this is already the case with Qubes 4.2.

**Comment by andrewdavidwong** [comment]
Since this is an enhancement request for the installer, and since this is already done for 4.2, this issue will be actionable only if 4.1 gets another patch release, which seems unlikely with 4.2 around the corner.

**Comment by rustybird** [comment]
Yeah it should probably be about qubes-dist-upgrade rather than the installer

**Comment by no-usernames-left** [comment]
> Yes, R4.2 has a new enough cryptsetup in the installer that defaults to `argon2id`.

What are the default memory cost, ops cost, and parallelism on 4.2 RC? Perhaps there is still some room for improvement here since the minimum system requirement for Qubes is significant enough to avoid causing problems by using i.e. 1024MiB memory cost.

**Comment by marmarek** [comment]
During installation you have a lot more RAM available than later during system runtime, because later you have several VMs running that occupy memory. I'd rather avoid situation where you need to shutdown everything just to unlock some partition...

**Comment by andrewdavidwong** [comment]
> Is it really the right choice to weaken the security of every user by watering down the default Argon2 settings just to make sure everyone can rekey LUKS with low free RAM?

Isn't this backwards? I was under the impression that you were suggesting strengthening the settings over and above the defaults. Aren't the current defaults in Qubes 4.2 just the unmodified defaults inherited from upstream?

Anyway, this is all off-topic for this issue. If you believe that Qubes should use stronger settings than the inherited upstream defaults, please open a separate issue for that.

**Comment by andrewdavidwong** [comment]
> I am indeed suggesting that. The wording I used was confusing; my apologies.

Ah, I see what you meant. No worries.

> I don't know, but they probably shouldn't be given the emphasis on high security in Qubes OS.

But LUKS is also security software, so there should also be an emphasis on high security in LUKS. (If there is not, then we shouldn't use it, and neither should anyone else. And yet, LUKS is the standard.) If LUKS doesn't ship with secure defaults, then a bug report should be submitted to the LUKS devs.

I think it's a mistake to assume that we should reject the upstream defaults of a piece of security software just because we also care about security. The mere fact that they're the defaults doesn't automatically make them insecure or unsuitable for us. The LUKS devs probably know what they're doing and have good reasons for selecting those defaults. We should reject the defaults only if we have specific, well-informed reasons to do so backed by some kind of evidence or expert advice.

For example, if you ask the LUKS devs about this, and they say that certain default parameters were chosen to accommodate less powerful devices and that all devices capable of running Qubes OS will work with more demanding parameters that would provide a real-world security benefit, then we certainly should consider making those more demanding parameters the defaults in Qubes OS. However, for all I know, they might instead say that this is already covered by the default `iter-time` parameter, which inherently adapts to more powerful hardware, since it's a function of the speed of the machine on which it is run. (This is just an example. I am not a cryptographer and have no idea what the LUKS devs will say. Please don't get caught up in the details of this specific example.)

So, my suggestion would be to try asking the LUKS devs (or, if they're unavailable, perhaps other cryptographers) for their advice and/or gather some relevant evidence. Then, if your findings indicate that it would actually be beneficial to change the default parameters for Qubes OS (e.g., there will be a real-world security gain that won't break things for any Qubes users), open an issue presenting your findings and requesting specific parameter changes.

**Comment by andrewdavidwong** [comment]
@no-usernames-left: Okay, then please open a separate issue with for this. (This issue is only about changing the KDF for in-place upgrades from 4.1 to 4.2, whereas changing the default Argon2id parameters will also affect fresh 4.2 installations, for example.)

**Comment by andrewdavidwong** [comment]
@no-usernames-left: No, you've misunderstood. There are two logically separate things here:

1. Changing the KDF to Argon2id.
2. Changing the Argon2id parameters.

This issue is about (1), because Qubes 4.1 and Qubes 4.2 use different KDFs by default, because they use different LUKS versions by default. Someone who upgrades in-place from Qubes 4.1 to 4.2 will still have the old KDF (unless this issue is completed).

The other issue you just opened (#8600) is about (2), because changing the Argon2id parameters is applicable only when the KDF is Argon2id. Issue #8600 could also include changing the Argon2id parameters in the case of in-place upgrades, but that's impossible until this issue is completed, because you can't turn the knobs until the knobs exist to be turned.
